api/src/main/java/jakarta/data/metamodel/restrict/CompositeRestrictionRecord.java

```diff
@@ -38,10 +40,12 @@ record CompositeRestrictionRecord<T>(
             throw new IllegalArgumentException(
                     "Cannot create a composite restriction without any restrictions to combine.");
         }
+        Objects.requireNonNull(type, "type must not be null");
+        restrictions.forEach(r -> Objects.requireNonNull(r, "Restriction must not be null"));
     }
 
     CompositeRestrictionRecord(Type type, List<Restriction<T>> restrictions) {
-        this(type, restrictions, false);
+        this(type, new ArrayList<>(restrictions), false);
```

Comment:
I did a copy of the List to avoid modification later as reference.

Reply:
Good catch. But in fact this is not enough. We need to also call `unmodifiableList()` everywhere the record is instantiated..

Reply:
The lack of immutable collections in Java is a massive hole in the whole design of `record` types :-(

Reply:
Actually, @otaviojava, if you check Nathan's latest code, none of this is necessary. All code paths which instantiate `CompositeRestrictionRecord` do in fact call `List.of()`.

Reply:
(Though there were related problems with `in()`, which I have fixed in #1075.)

Reply:
Hum, let me rebase this PR as well.

Reply:
You were right, removing it.